### PR TITLE
8344328: (dc) DatagramChannelImpl.blockingReceive can now synchronize on packet

### PR DESCRIPTION
**Pull request by AlanBateman**
Now that JEP 491 is integrated, we can restore long-standing behavior in the DatagramSocket adaptor to synchronize on the DatagramPacket for the duration of the send/receive.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8344328](https://bugs.openjdk.org/browse/JDK-8344328): (dc) DatagramChannelImpl.blockingReceive can now synchronize on packet (**Enhancement** - P4)

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) Review applies to [a50fdddc](https://git.openjdk.org/jdk/pull/22212/files/a50fdddc670413a3c41be9efe095d7011dfeb11e)
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22212/head:pull/22212` \
`$ git checkout pull/22212`

Update a local copy of the PR: \
`$ git checkout pull/22212` \
`$ git pull https://git.openjdk.org/jdk.git pull/22212/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22212`

View PR using the GUI difftool: \
`$ git pr show -t 22212`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22212.diff">https://git.openjdk.org/jdk/pull/22212.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22212#issuecomment-2484935888)
</details>

### CODE REVIEW

File: src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java [hunk 1]
```diff
@@ -843,38 +839,36 @@ void blockingSend(DatagramPacket p) throws IOException {
...
                         target = (InetSocketAddress) p.getSocketAddress();
                     }
-                }
-
-                // send the datagram (does not block)
-                send(src, target);
 
-            } finally {
-                if (src != null) Util.offerFirstTemporaryDirectBuffer(src);
+                    // send the datagram (does not block)
```
Comment (liach) [inline src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java:865]: Now this send is in the synchronized block; should this "does not block" be moved to the top level synchronized block?

Comment (AlanBateman) [reply src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java:865]: No, it's in the correct place.

**Review by jaikiran — APPROVED** [review]
This looks reasonable to me. I believe the previous decision to reduce the scope of `synchronized` blocks in this code was done for https://bugs.openjdk.org/browse/JDK-8312166.

**Comment by AlanBateman** [comment]
> This looks reasonable to me. I believe the previous decision to reduce the scope of `synchronized` blocks in this code was done for https://bugs.openjdk.org/browse/JDK-8312166.

That's right, I should have linked to that change in JDK 22. The removal of the SecurityManager support, and the follow-up work in JDK-8344112, means the changes to hold the lock for the duration of both I/O ops is much simpler now as a lot of complexity and code paths go away.

**Review by dfuch — COMMENTED** [review]
Given that anyone can synchronize on DatagramPacket, would it be better to synchronize on `p` in the DatagramSocketAdaptor, before obtaining the readLock/writeLock, in order to reduce the chances of potentially triggering a deadlock?

**Comment by AlanBateman** [comment]
> Given that anyone can synchronize on DatagramPacket, would it be better to synchronize on `p` in the DatagramSocketAdaptor, before obtaining the readLock/writeLock, in order to reduce the chances of potentially triggering a deadlock?

DatagramPacket isn't specified to be tread safe but as with many JDK 1.0 era APIs, all methods are synchronized. I think your question is about lock ordering as I/O operations require a read or write lock. This isn't specified anywhere but the implementation works consistently to avoid deadlock when using the APIs. If something synchronizes on a DatagramPacket then it would delay, maybe indefinitely, a send or receive op. So I think your concern is whether it might  delay, maybe indefinitely, other send or receive operations with other datagram packages, right? I don't mind changing it. Arguably this would give us closer to what was there before JEP 373 although the legacy implementation didn't correctly coordinate concurrently usage beyond that.

**Comment by dfuch** [comment]
> DatagramPacket isn't specified to be tread safe but as with many JDK 1.0 era APIs, all methods are synchronized. I think your question is about lock ordering as I/O operations require a read or write lock. This isn't specified anywhere but the implementation works consistently to avoid deadlock when using the APIs. If something synchronizes on a DatagramPacket then it would delay, maybe indefinitely, a send or receive op. So I think your concern is whether it might delay, maybe indefinitely, other send or receive operations with other datagram packages, right? I don't mind changing it. Arguably this would give us closer to what was there before JEP 373 although the legacy implementation didn't correctly coordinate concurrently usage beyond that.

Correct - my concern is with lock ordering, and existing custom code that might have been written before JEP 373, and that might synchronize on DatagramPacket. I'm OK if you think it's better to keep all the synchro in DatagramChannelImpl, but since the methods that take a DatagramPacket are only called from the DatagramSocketAdaptor I was wondering if it would be more pertinent to synchronize on DatagramPacket from there (and bring it closer to pre JEP 373 implementation).

**Comment by AlanBateman** [comment]
> Correct - my concern is with lock ordering, and existing custom code that might have been written before JEP 373, and that might synchronize on DatagramPacket. I'm OK if you think it's better to keep all the synchro in DatagramChannelImpl, but since the methods that take a DatagramPacket are only called from the DatagramSocketAdaptor I was wondering if it would be more pertinent to synchronize on DatagramPacket from there (and bring it closer to pre JEP 373 implementation).

The read/write lock used by DatagramChannel is not exposed so the ordering is hard to observe. In any case, I think what we have in the PR aligns with the locking that has been there since JDK 1.0. That is, it does in synchronized (p) in DatagramSocketAdaptor.send/receive It's a bit icky as the blockingXXX methods needs to assert that the current thread holds the lock, and doing in blockingXXX would work the same (the difference is not observable).

**Comment by dfuch** [comment]
> The read/write lock used by DatagramChannel is not exposed so the ordering is hard to observe. In any case, I think what we have in the PR aligns with the locking that has been there since JDK 1.0. That is, it does in synchronized (p) in DatagramSocketAdaptor.send/receive It's a bit icky as the blockingXXX methods needs to assert that the current thread holds the lock, and doing in blockingXXX would work the same (the difference is not observable).

I see. I'm OK with both solutions - so choose the one you like more and I will (re-)approve if needed.

**Review by dfuch — APPROVED** [review]

**Comment by AlanBateman** [comment]
Thanks, let's go with what we have.

**Pull request closed without merging**